py/gc.c

```diff
@@ -803,15 +803,34 @@ void gc_dump_info(void) {
            (uint)info.num_1block, (uint)info.num_2block, (uint)info.max_block, (uint)info.max_free);
 }
 
+#define DUMP_BLOCK_INFO (MICROPY_GC_DUMP_ALLOC_TABLE_MODE & MICROPY_GC_DUMP_ALLOC_TABLE_MODE_BLOCKS)
+#define DUMP_FRAGMENTATION_INFO (MICROPY_GC_DUMP_ALLOC_TABLE_MODE & MICROPY_GC_DUMP_ALLOC_TABLE_MODE_FRAG)
+#define NUM_FRAGMENTATION_CLASSES (20)
+
+#if DUMP_FRAGMENTATION_INFO
+STATIC void gc_update_fragmentation_stats(size_t n, size_t *classes) {
+    for (size_t c = 1, i = 0; c < n; c <<= 1, ++i) {
+        classes[i] += n / c;
+    }
+}
+#endif
+
 void gc_dump_alloc_table(void) {
     GC_ENTER();
+    #if DUMP_BLOCK_INFO
     static const size_t DUMP_BYTES_PER_LINE = 64;
+    #endif
     #if !EXTENSIVE_HEAP_PROFILING
     // When comparing heap output we don't want to print the starting
     // pointer of the heap because it changes from run to run.
     mp_printf(&mp_plat_print, "GC memory layout; from %p:", MP_STATE_MEM(gc_pool_start));
     #endif
+    #if DUMP_FRAGMENTATION_INFO
+    size_t n = 0;
+    size_t classes[NUM_FRAGMENTATION_CLASSES] = {0};
+    #endif
     for (size_t bl = 0; bl < MP_STATE_MEM(gc_alloc_table_byte_len) * BLOCKS_PER_ATB; bl++) {
+        #if DUMP_BLOCK_INFO
```

Comment:
So if DUMP_BLOCK_INFO nor DUMP_FRAGMENTATION_INFO is set, should this function run at all?

Reply:
I guess not. I've #if'd it in modmicropython.c